pretyping/pretyping.mli

```diff
@@ -32,12 +32,17 @@ val clear_bidirectionality_hint : GlobRef.t -> unit
 
 (** An auxiliary function for searching for fixpoint guard indexes *)
 
+type possible_guard =
+  bool (* true if can be a cofixpoint *) *
+  int list list (* list of non-empty lists of possible indices if can be a fixpoint *)
+
 val search_guard :
-  ?loc:Loc.t -> ?evars:CClosure.evar_handler -> env -> int list list -> Constr.rec_declaration -> int array
+  ?loc:Loc.t -> ?evars:CClosure.evar_handler -> env ->
+  possible_guard -> Constr.rec_declaration -> int array option
```

Comment:
I guess the option means None -> cofix, Some -> fix
This should be in a comment
Or maybe the maybe_cofix bool in possible_guard should be a GADT so that we don't have to Option.get when we know we want a fixpoint.

Reply:
GADT: I'll think also about it.

Reply:
There are two cases in the code when we know in advance that it is a fixpoint. For these cases, maybe could we just provide a variant of `search_guard` in `pretyping.ml` that takes only a `int list list` and returns necessarily an `int array`? That would avoid the caller of `search_guard` to know about the invariants of `pretyping`. I'm tempted to go this direction if ok for you.